File: src/skills/manager.zig
const std = @import("std");
const loader = @import("loader.zig");

pub const SkillManager = struct {
    allocator: std.mem.Allocator,
    io: std.Io,
    skill_loader: *loader.SkillLoader,
    skills_dir: []const u8,

    pub fn dummyHandler(alloc: std.mem.Allocator, io_arg: std.Io, args: std.json.Value) anyerror![]const u8 {
        _ = alloc;
        _ = io_arg;
        _ = args;
        return "not dispatched via registry";
    }

    pub fn init(allocator: std.mem.Allocator, io: std.Io, skill_loader: *loader.SkillLoader, skills_dir: []const u8) SkillManager {
        return .{
            .allocator = allocator,
            .io = io,
            .skill_loader = skill_loader,
            .skills_dir = skills_dir,
        };
    }

    fn copyDirRecursive(self: *SkillManager, src_dir: []const u8, dest_dir: []const u8) !usize {
        var file_count: usize = 0;
        const cwd = std.Io.Dir.cwd();

        var src = cwd.openDir(self.io, src_dir, .{ .iterate = true }) catch return 0;
        defer src.close(self.io);

        var walker = src.walk(self.allocator) catch return 0;
        defer walker.deinit();

        while (try walker.next(self.io)) |entry| {
            if (entry.kind == .file) {
                const src_path = try std.fs.path.join(self.allocator, &.{ src_dir, entry.path });
                defer self.allocator.free(src_path);

                const dest_path = try std.fs.path.join(self.allocator, &.{ dest_dir, entry.path });
                defer self.allocator.free(dest_path);

                const content = cwd.readFileAlloc(self.io, src_path, self.allocator, std.Io.Limit.limited(10 * 1024 * 1024)) catch continue;
                defer self.allocator.free(content);

                const parent = std.fs.path.dirname(dest_path);
                if (parent) |p| {
                    cwd.createDirPath(self.io, p) catch {};
                }

                const file = try cwd.createFile(self.io, dest_path, .{});
                defer file.close(self.io);
                try file.writeStreamingAll(self.io, content);

                file_count += 1;
            }
        }

        return file_count;
    }

    pub fn skillsList(self: *SkillManager, args: std.json.Value) ![]const u8 {
        const all_skills = self.skill_loader.loadAll() catch return "Error: failed to load skills";
        defer {
            for (all_skills) |s| {
                self.allocator.free(s.name);
                self.allocator.free(s.description);
                self.allocator.free(s.body);
                self.allocator.free(s.path);
                self.allocator.free(s.version);
            }
            self.allocator.free(all_skills);
        }

        const filtered_skills: []const loader.Skill = all_skills;
        var category_filter: ?[]const u8 = null;
        if (args.object.get("category")) |cat| {
            category_filter = cat.string;
        }

        var result: std.ArrayList(u8) = .empty;
        var count: usize = 0;

        for (filtered_skills) |s| {
            if (category_filter) |cf| {
                if (std.mem.indexOf(u8, s.path, cf) == null) continue;
            }
            const desc = if (s.description.len > 200) s.description[0..200] else s.description;
            try result.print(self.allocator, "  {s}: {s} (v{s})\n", .{ s.name, desc, s.version });
            count += 1;
        }

        if (count == 0) return "No skills found.";
        return try result.toOwnedSlice(self.allocator);
    }

    pub fn skillView(self: *SkillManager, args: std.json.Value) ![]const u8 {
        const name = if (args.object.get("name")) |n| n.string else return "Error: name is required";

        const skill = try self.skill_loader.findSkill(name) orelse
            return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' not found", .{name});

        if (args.object.get("file_path")) |fp| {
            const target_path = try std.fs.path.join(self.allocator, &.{ skill.path, fp.string });
            defer self.allocator.free(target_path);

            const cwd = std.Io.Dir.cwd();
            return cwd.readFileAlloc(self.io, target_path, self.allocator, std.Io.Limit.limited(1024 * 1024)) catch
                try std.fmt.allocPrint(self.allocator, "Error: file '{s}' not found in skill '{s}'", .{ fp.string, name });
        }

        return try self.skill_loader.viewSkillFile(skill);
    }

    pub fn skillManage(self: *SkillManager, args: std.json.Value) ![]const u8 {
        const cwd = std.Io.Dir.cwd();
        const action_val = args.object.get("action") orelse return "Error: action is required";
        const action = action_val.string;
        const name = if (args.object.get("name")) |n| n.string else return "Error: name is required";

        if (std.mem.eql(u8, action, "create")) {
            const content = if (args.object.get("content")) |c| c.string else return "Error: content is required for create";
            if (!std.mem.startsWith(u8, content, "---")) return "Error: SKILL.md must start with YAML frontmatter (---)";

            var skill_dir_buf: std.ArrayList(u8) = .empty;
            if (args.object.get("category")) |cat| {
                try skill_dir_buf.print(self.allocator, "{s}/{s}/{s}", .{ self.skills_dir, cat.string, name });
            } else {
                try skill_dir_buf.print(self.allocator, "{s}/{s}", .{ self.skills_dir, name });
            }

            const skill_md_path = try std.fs.path.join(self.allocator, &.{ skill_dir_buf.items, "SKILL.md" });
            const existing = cwd.readFileAlloc(self.io, skill_md_path, self.allocator, std.Io.Limit.limited(1)) catch null;
            if (existing) |_| {
                self.allocator.free(skill_md_path);
                return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' already exists. Use 'patch' or 'edit'.", .{name});
            }

            cwd.createDirPath(self.io, skill_dir_buf.items) catch {};

            const file = try cwd.createFile(self.io, skill_md_path, .{});
            defer file.close(self.io);
            try file.writeStreamingAll(self.io, content);

            return try std.fmt.allocPrint(self.allocator, "Skill '{s}' created at {s}", .{ name, skill_dir_buf.items });
        }

        if (std.mem.eql(u8, action, "patch")) {
            const old_string = if (args.object.get("old_string")) |o| o.string else return "Error: old_string is required for patch";
            const new_string_val = args.object.get("new_string") orelse return "Error: new_string is required for patch";
            const new_string = new_string_val.string;
            const replace_all = if (args.object.get("replace_all")) |ra| ra.bool else false;

            const skill = try self.skill_loader.findSkill(name) orelse
                return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' not found", .{name});

            var target_path: []const u8 = undefined;
            if (args.object.get("file_path")) |fp| {
                target_path = try std.fs.path.join(self.allocator, &.{ skill.path, fp.string });
            } else {
                target_path = try std.fs.path.join(self.allocator, &.{ skill.path, "SKILL.md" });
            }
            defer self.allocator.free(target_path);

            const current = cwd.readFileAlloc(self.io, target_path, self.allocator, std.Io.Limit.limited(1024 * 1024)) catch
                return "Error: failed to read skill file";
            defer self.allocator.free(current);

            const count = std.mem.count(u8, current, old_string);
            if (count == 0) return "Error: old_string not found";
            if (count > 1 and !replace_all)
                return try std.fmt.allocPrint(self.allocator, "Error: old_string found {d} times. Use replace_all=true or add more context.", .{count});

            var new_content: std.ArrayList(u8) = .empty;
            if (replace_all) {
                var parts = std.mem.splitSequence(u8, current, old_string);
                var first = true;
                while (parts.next()) |part| {
                    if (!first) try new_content.appendSlice(self.allocator, new_string);
                    try new_content.appendSlice(self.allocator, part);
                    first = false;
                }
            } else {
                if (std.mem.indexOf(u8, current, old_string)) |idx| {
                    try new_content.appendSlice(self.allocator, current[0..idx]);
                    try new_content.appendSlice(self.allocator, new_string);
                    try new_content.appendSlice(self.allocator, current[idx + old_string.len ..]);
                }
            }

            const file = try cwd.createFile(self.io, target_path, .{});
            defer file.close(self.io);
            try file.writeStreamingAll(self.io, new_content.items);

            return try std.fmt.allocPrint(self.allocator, "Patched '{s}': {d} replacement(s)", .{ name, count });
        }

        if (std.mem.eql(u8, action, "edit")) {
            const content = if (args.object.get("content")) |c| c.string else return "Error: content is required for edit";
            const skill = try self.skill_loader.findSkill(name) orelse
                return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' not found", .{name});
            const skill_md_path = try std.fs.path.join(self.allocator, &.{ skill.path, "SKILL.md" });
            defer self.allocator.free(skill_md_path);
            const file = try cwd.createFile(self.io, skill_md_path, .{});
            defer file.close(self.io);
            try file.writeStreamingAll(self.io, content);
            return try std.fmt.allocPrint(self.allocator, "Skill '{s}' fully rewritten", .{name});
        }

        if (std.mem.eql(u8, action, "delete")) {
            const skill = try self.skill_loader.findSkill(name) orelse
                return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' not found", .{name});
            cwd.deleteTree(self.io, skill.path) catch return "Error: failed to delete skill";
            return try std.fmt.allocPrint(self.allocator, "Skill '{s}' deleted", .{name});
        }

        if (std.mem.eql(u8, action, "write_file")) {
            const file_path = if (args.object.get("file_path")) |fp| fp.string else return "Error: file_path is required for write_file";
            const file_content = if (args.object.get("file_content")) |fc| fc.string else return "Error: file_content is required for write_file";

            const allowed_dirs = [_][]const u8{ "references", "templates", "scripts", "assets" };
            var first_dir: []const u8 = "";
            if (std.mem.indexOf(u8, file_path, "/")) |slash_pos| {
                first_dir = file_path[0..slash_pos];
            } else {
                first_dir = file_path;
            }
            var is_allowed = false;
            for (allowed_dirs) |ad| {
                if (std.mem.eql(u8, first_dir, ad)) {
                    is_allowed = true;
                    break;
                }
            }
            if (!is_allowed)
                return "Error: file must be under one of: references, templates, scripts, assets";

            const skill = try self.skill_loader.findSkill(name) orelse
                return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' not found", .{name});

            const target_path = try std.fs.path.join(self.allocator, &.{ skill.path, file_path });
            defer self.allocator.free(target_path);

            const parent_dir = std.fs.path.dirname(target_path) orelse target_path;
            cwd.createDirPath(self.io, parent_dir) catch {};

            const file = try cwd.createFile(self.io, target_path, .{});
            defer file.close(self.io);
            try file.writeStreamingAll(self.io, file_content);

            return try std.fmt.allocPrint(self.allocator, "Written {s} in skill '{s}'", .{ file_path, name });
        }

        if (std.mem.eql(u8, action, "remove_file")) {
            const file_path = if (args.object.get("file_path")) |fp| fp.string else return "Error: file_path is required for remove_file";

            const skill = try self.skill_loader.findSkill(name) orelse
                return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' not found", .{name});

            const target_path = try std.fs.path.join(self.allocator, &.{ skill.path, file_path });
            defer self.allocator.free(target_path);

            cwd.deleteFile(self.io, target_path) catch
                return try std.fmt.allocPrint(self.allocator, "Error: file '{s}' not found", .{file_path});

            return try std.fmt.allocPrint(self.allocator, "Removed {s} from skill '{s}'", .{ file_path, name });
        }

        if (std.mem.eql(u8, action, "install")) {
            const source_dir = if (args.object.get("source_dir")) |sd| sd.string else return "Error: source_dir is required for install";

            const src_base = "src/skills";
            var src_path_buf: std.ArrayList(u8) = .empty;
            try src_path_buf.print(self.allocator, "{s}/{s}", .{ src_base, source_dir });
            defer src_path_buf.deinit(self.allocator);

            const src_exists = cwd.openDir(self.io, src_path_buf.items, .{}) catch null;
            if (src_exists == null) {
                return try std.fmt.allocPrint(self.allocator, "Error: source directory '{s}/{s}' not found", .{ src_base, source_dir });
            } else {
                src_exists.?.close(self.io);
            }

            var dest_path_buf: std.ArrayList(u8) = .empty;
            try dest_path_buf.print(self.allocator, "{s}/{s}", .{ self.skills_dir, name });
            defer dest_path_buf.deinit(self.allocator);

            const existing = cwd.openDir(self.io, dest_path_buf.items, .{}) catch null;
            if (existing != null) {
                existing.?.close(self.io);
                return try std.fmt.allocPrint(self.allocator, "Error: skill '{s}' already installed. Use 'patch' or 'edit' to modify.", .{name});
            }

            cwd.createDirPath(self.io, dest_path_buf.items) catch {};

            const copied = try self.copyDirRecursive(src_path_buf.items, dest_path_buf.items);
            if (copied > 0) {
                return try std.fmt.allocPrint(self.allocator, "Skill '{s}' installed from '{s}' ({d} files copied)", .{ name, source_dir, copied });
            } else {
                cwd.deleteTree(self.io, dest_path_buf.items) catch {};
                return try std.fmt.allocPrint(self.allocator, "Error: no files found in source directory '{s}'", .{source_dir});
            }
        }

        return try std.fmt.allocPrint(self.allocator, "Error: unknown action '{s}'. Use: create, patch, edit, delete, write_file, remove_file, install", .{action});
    }
};

File: py/skills/manager.py
"""
Skill Manager (Chapter 10-12)

skill_manage tool: create, patch, edit, delete.
skills_list and skill_view tools.
"""

import json
import yaml
from pathlib import Path
from py.tool_registry import registry

# Set by cli.py
_skill_loader = None
_skills_dir = None


def set_skill_loader(loader, skills_dir):
    global _skill_loader, _skills_dir
    _skill_loader = loader
    _skills_dir = skills_dir


def skills_list(category: str = None) -> str:
    """List all skills with metadata (progressive disclosure tier 1)."""
    if _skill_loader is None:
        return "Error: skills not initialized"
    all_skills = _skill_loader.load_all()
    if category:
        all_skills = [s for s in all_skills if category in str(s["path"])]
    result = []
    for s in all_skills:
        result.append({
            "name": s["name"],
            "description": s["description"][:200],
            "version": s.get("version", "1.0.0"),
        })
    return json.dumps(result, indent=2) if result else "No skills found."


def skill_view(name: str, file_path: str = None) -> str:
    """Load full skill content (progressive disclosure tier 2-3)."""
    if _skill_loader is None:
        return "Error: skills not initialized"
    skill = _skill_loader.find_skill(name)
    if not skill:
        return f"Error: skill '{name}' not found"
    if file_path:
        target = skill["path"] / file_path
        if not target.exists():
            return f"Error: file '{file_path}' not found in skill '{name}'"
        return target.read_text()
    return (skill["path"] / "SKILL.md").read_text()


def skill_manage(action: str, name: str, content: str = None,
                 category: str = None, old_string: str = None,
                 new_string: str = None, replace_all: bool = False,
                 file_path: str = None, file_content: str = None) -> str:
    """Manage skills: create, edit, patch, delete, write_file, remove_file."""
    if _skills_dir is None:
        return "Error: skills not initialized"

    if action == "create":
        if not content:
            return "Error: content is required for create (full SKILL.md text)"
        # Validate frontmatter
        if not content.startswith("---"):
            return "Error: SKILL.md must start with YAML frontmatter (---)"
        # Build path
        if category:
            skill_dir = _skills_dir / category / name
        else:
            skill_dir = _skills_dir / name
        if (skill_dir / "SKILL.md").exists():
            return f"Error: skill '{name}' already exists. Use 'patch' or 'edit'."
        skill_dir.mkdir(parents=True, exist_ok=True)
        (skill_dir / "SKILL.md").write_text(content)
        return json.dumps({"success": True, "message": f"Skill '{name}' created at {skill_dir}"})

    elif action == "patch":
        if not old_string:
            return "Error: old_string is required for patch"
        if new_string is None:
            return "Error: new_string is required for patch"
        skill = _skill_loader.find_skill(name)
        if not skill:
            return f"Error: skill '{name}' not found"
        target = skill["path"] / (file_path or "SKILL.md")
        if not target.exists():
            return f"Error: file not found: {target}"
        current = target.read_text()
        count = current.count(old_string)
        if count == 0:
            return f"Error: old_string not found in {target.name}"
        if count > 1 and not replace_all:
            return f"Error: old_string found {count} times. Use replace_all=true or add more context."
        new_content = current.replace(old_string, new_string, -1 if replace_all else 1)
        target.write_text(new_content)
        return json.dumps({"success": True, "message": f"Patched '{name}': {count} replacement(s)"})

    elif action == "edit":
        if not content:
            return "Error: content is required for edit (full SKILL.md text)"
        skill = _skill_loader.find_skill(name)
        if not skill:
            return f"Error: skill '{name}' not found"
        (skill["path"] / "SKILL.md").write_text(content)
        return json.dumps({"success": True, "message": f"Skill '{name}' fully rewritten"})

    elif action == "delete":
        skill = _skill_loader.find_skill(name)
        if not skill:
            return f"Error: skill '{name}' not found"
        import shutil
        shutil.rmtree(skill["path"])
        return json.dumps({"success": True, "message": f"Skill '{name}' deleted"})

    elif action == "write_file":
        if not file_path or not file_content:
            return "Error: file_path and file_content required for write_file"
        skill = _skill_loader.find_skill(name)
        if not skill:
            return f"Error: skill '{name}' not found"
        allowed = {"references", "templates", "scripts", "assets"}
        first_dir = Path(file_path).parts[0] if Path(file_path).parts else ""
        if first_dir not in allowed:
            return f"Error: file must be under one of: {', '.join(sorted(allowed))}"
        target = skill["path"] / file_path
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(file_content)
        return json.dumps({"success": True, "message": f"Written {file_path} in skill '{name}'"})

    elif action == "remove_file":
        if not file_path:
            return "Error: file_path required for remove_file"
        skill = _skill_loader.find_skill(name)
        if not skill:
            return f"Error: skill '{name}' not found"
        target = skill["path"] / file_path
        if not target.exists():
            return f"Error: file not found: {file_path}"
        target.unlink()
        return json.dumps({"success": True, "message": f"Removed {file_path} from skill '{name}'"})

    else:
        return f"Error: unknown action '{action}'"


# Register tools
registry.register(
    name="skills_list",
    description="List all available skills with name and description (metadata only).",
    parameters={
        "type": "object",
        "properties": {
            "category": {"type": "string", "description": "Optional category filter"},
        },
    },
    handler=skills_list,
    category="skills",
)

registry.register(
    name="skill_view",
    description="View the full content of a skill or a specific supporting file within it.",
    parameters={
        "type": "object",
        "properties": {
            "name": {"type": "string", "description": "Skill name"},
            "file_path": {"type": "string", "description": "Optional: path to a supporting file (e.g. references/api.md)"},
        },
        "required": ["name"],
    },
    handler=skill_view,
    category="skills",
)

registry.register(
    name="skill_manage",
    description=(
        "Manage skills (create, update, delete). Skills are procedural memory.\n"
        "Actions: create (full SKILL.md), patch (old_string/new_string), "
        "edit (full rewrite), delete, write_file, remove_file.\n"
        "After difficult tasks, offer to save as a skill. Confirm before creating/deleting."
    ),
    parameters={
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": ["create", "patch", "edit", "delete", "write_file", "remove_file"],
            },
            "name": {"type": "string", "description": "Skill name (lowercase, hyphens)"},
            "content": {"type": "string", "description": "Full SKILL.md for create/edit"},
            "category": {"type": "string", "description": "Category subdirectory for create"},
            "old_string": {"type": "string", "description": "Text to find (patch)"},
            "new_string": {"type": "string", "description": "Replacement text (patch)"},
            "replace_all": {"type": "boolean", "description": "Replace all occurrences (patch)"},
            "file_path": {"type": "string", "description": "Supporting file path"},
            "file_content": {"type": "string", "description": "Content for write_file"},
        },
        "required": ["action", "name"],
    },
    handler=skill_manage,
    category="skills",
)
